examples/tp+pp/pippy_tp.py

```diff
@@ -0,0 +1,214 @@
+import argparse
+import os
+import unittest
+
+import torch
+
+import pippy.fx
+from pippy import run_pippy
+from pippy.IR import (
+    Pipe,
+    pipe_split,
+)
+from pippy.PipelineDriver import (
+    PipelineDriverFillDrain,
+    PipelineDriver1F1B,
+    PipelineDriverBase,
+    PipelineDriverInterleaved1F1B,
+)
+
+from torch.distributed._tensor import (
+    DeviceMesh,
+)
+from torch.distributed.tensor.parallel import (
+    PairwiseParallel,
+    parallelize_module,
+)
+
+
+schedules = {
+    "FillDrain": PipelineDriverFillDrain,
+    "1F1B": PipelineDriver1F1B,
+    "Interleaved1F1B": PipelineDriverInterleaved1F1B,
+}
+
+pippy.fx.Tracer.proxy_buffer_attributes = True
+
+
+class MLPModule(torch.nn.Module):
+    def __init__(self, d_hid):
+        super(MLPModule, self).__init__()
+        self.net1 = torch.nn.Linear(d_hid, d_hid)
+        self.relu = torch.nn.ReLU()
+        self.net2 = torch.nn.Linear(d_hid, d_hid)
+
+    def forward(self, x):
+        x = self.net1(x)
+        x = self.relu(x)
+        x = self.net2(x)
+        return x
+
+
+class ExampleCode(torch.nn.Module):
+    def __init__(self, d_hid):
+        super().__init__()
+        self.mlp0 = MLPModule(d_hid)
+        self.mlp1 = MLPModule(d_hid)
+        self.mlp2 = MLPModule(d_hid)
+        self.mlp3 = MLPModule(d_hid)
+
+    def forward(self, x):
+        x = self.mlp0(x)
+        pipe_split()
+        x = self.mlp1(x)
+        pipe_split()
+        x = self.mlp2(x)
+        pipe_split()
+        x = self.mlp3(x)
+        return x
+
+
+def run_gspmd(pp_ranks, args):
+    chunks = args.pp_group_size
+    device_type = "cuda" if args.cuda else "cpu"
+
+    # Figure out my PP rank
+    pp_rank = args.rank // args.tp_group_size
+    print(f"GSPMD rank {args.rank}, PP ranks: {pp_ranks}, my rank in pipe: {pp_rank}")
+
+    d_hid = 256
+    batch_size_per_chunk = 8
+    inp_size = [chunks * batch_size_per_chunk, d_hid]
+
+    # Ensure all tp ranks have same input.
+    torch.manual_seed(0)
+    inp = torch.rand(*inp_size, device=device_type)
+
+    """
+    # Reference run
+    # `torchrun --nproc_per_node=2 pippy_tp_mlp.py --world_size=2 --pp_group_size=1`
+    ec_tp = ExampleCode(d_hid)
+    ec_tp.to(args.device)
+
+    start_idx = 0
+    device_mesh = DeviceMesh(
+        device_type,
+        list(range(start_idx, start_idx + args.tp_group_size)),
+    )
+    print(f"Rank {args.rank} calling parallelize_module with {device_mesh}")
+    parallelize_module(ec_tp, device_mesh, PairwiseParallel())
+    print(f"Rank {args.rank} sharding complete")
+
+    ref_out = ec_tp(inp)
+    print(f"Ref out: {ref_out.size()}")
+    """
+
+    # PiPPy run
+    ec = ExampleCode(d_hid)
+    ec.to(args.device)
+
+    # PiPPy tracing
+    ec_pipe = Pipe.from_tracing(ec)
+    if args.rank == 0:
+        print(ec_pipe.split_gm)
+
+    # Ask PiPPy to export stage submodule
+    submod = ec_pipe.export(stage_id=pp_rank)
+
+    # Create TP device mesh
+    my_device_mesh = None
```

Comment:
This should be using a single global 2D DeviceMesh with one dimension assigned for PP and the other for TP.

Reply:
Great callout @kumpera.
It requires PiPPy to support `DeviceMesh` first. I will look into that soon.